### ISSUE TITLE
Implement Catalogue Search

### ISSUE THREAD

**Issue by nicholascar** [RDFLib/prez-ui#58]
CatPrez' top-level page should be a search page, like SpacePrez' is.

e.g. https://data.idnau.org/c should look like https://data.idnau.org/s with an altered emphasis, as below

![Image](https://github.com/RDFLib/prez-ui/assets/7321872/87c53f4a-dcd0-4296-b677-bbce88f9b99f)

**Comment by nicholascar** [comment]
Query for "Catalogues to include":

```
PREFIX dcterms: <http://purl.org/dc/terms/>
PREFIX dcat: <http://www.w3.org/ns/dcat#>

SELECT ?c ?t
WHERE {
  GRAPH ?g {
  ?c a dcat:Catalog ;
  	dcterms:title ?t .
  }
}
ORDER BY ?t
```

**Comment by nicholascar** [comment]
Query by text input using weightings as:

* exact title match
* partial title match
* partial description match

```
PREFIX dcterms: <http://purl.org/dc/terms/>
PREFIX dcat: <http://www.w3.org/ns/dcat#>

SELECT DISTINCT ?r ?t ?d (SUM(?w) AS ?weight)
WHERE {
  ?r a dcat:Resource ;
    ^dcterms:hasPart <https://data.idnau.org/pid/democat>  ;   
  .
  {
    ?r
      dcterms:title ?t ;
      dcterms:description ?d ;        
    .
    BIND (50 AS ?w)
    FILTER REGEX(?t, "^Indigenous$", "i")
  }
  UNION
  {
    ?r 
      dcterms:title ?t ;
      dcterms:description ?d ;      
    .
    BIND (10 AS ?w)
    FILTER REGEX(?t, "Indigenous", "i")
  }
  UNION
  {
    ?r 
      dcterms:title ?t ;
      dcterms:description ?d ;      
    .
    BIND (5 AS ?w)
    FILTER REGEX(?d, "Indigenous", "i")
  }  
}
GROUP BY ?r ?t ?d ?match
ORDER BY DESC(?weight) ?t
```

**Comment by nicholascar** [comment]
Query for "Theme filter" the retrieves the top 4 most commonly used themes for Resources in the `https://data.idnau.org/pid/democat` catalogue:

```
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX dcterms: <http://purl.org/dc/terms/>
PREFIX dcat: <http://www.w3.org/ns/dcat#>

SELECT ?th ?pl (COUNT(?th) AS ?count)
WHERE {
  ?r a dcat:Resource ;
    ^dcterms:hasPart <https://data.idnau.org/pid/democat>  ;   
  .
  
  ?r dcat:theme ?th .
  ?th skos:prefLabel ?pl .
}
GROUP BY ?th ?pl
ORDER BY DESC(?count) ?pl 
LIMIT 4
```

**Comment by nicholascar** [comment]
Query for "Spatial filter" with a given geometry:

```
PREFIX dcat: <http://www.w3.org/ns/dcat#>
PREFIX dcterms: <http://purl.org/dc/terms/>
PREFIX geo: <http://www.opengis.net/ont/geosparql#>
PREFIX geof: <http://www.opengis.net/def/function/geosparql/>

SELECT *
WHERE {
  GRAPH ?g {
    ?r a dcat:Resource ;
      ^dcterms:hasPart <https://data.idnau.org/pid/democat>  ;  
      geo:hasBoundingBox/geo:asWKT ?wkt ;
    .
    
    FILTER (geof:sfOverlaps("POLYGON ((159.86557837137977 -11.168170683361419, 159.86557837137977 -39.83427815914998, 111.08628149637977 -39.83427815914998, 111.08628149637977 -11.168170683361419, 159.86557837137977 -11.168170683361419)))"^^geo:wktLiteral, ?wkt))
  }
}
```

**Comment by nicholascar** [comment]
Combined query using all filters:

```
PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
PREFIX dcat: <http://www.w3.org/ns/dcat#>
PREFIX dcterms: <http://purl.org/dc/terms/>
PREFIX geo: <http://www.opengis.net/ont/geosparql#>
PREFIX geof: <http://www.opengis.net/def/function/geosparql/>

SELECT ?r ?t ?d ?th ?thpl ?weight
WHERE {
  GRAPH ?g {
    # Only look for Resources (not spatial Datasets)
    ?r a dcat:Resource .

    # Only look in DemoCat
    ?r  ^dcterms:hasPart <https://data.idnau.org/pid/democat>  .
    
    # Weighted text search
    {
      SELECT DISTINCT ?r ?t ?d (SUM(?w) AS ?weight)
      WHERE {
        ?r a dcat:Resource .
        {
          ?r
            dcterms:title ?t ;
            dcterms:description ?d ;
          .
          BIND (50 AS ?w)
          FILTER REGEX(?t, "^Police$", "i")
        }
        UNION
        {
          ?r 
            dcterms:title ?t ;
            dcterms:description ?d ;
          .
          BIND (10 AS ?w)
          FILTER REGEX(?t, "Police", "i")
        }
        UNION
        {
          ?r 
            dcterms:title ?t ;
            dcterms:description ?d ;
          .
          BIND (5 AS ?w)
          FILTER REGEX(?d, "Police", "i")
        }  
      }
      GROUP BY ?r ?t ?d ?match
      ORDER BY DESC(?weight) ?t
    }
    
    # theme filter
    ?r dcat:theme <https://vocabularyserver.com/apais/xml.php?skosTema=28> .
    <https://vocabularyserver.com/apais/xml.php?skosTema=28> skos:prefLabel ?thpl .
    
    # Spatial Filter
    ?r geo:hasBoundingBox/geo:asWKT ?wkt ;
    FILTER (geof:sfOverlaps("POLYGON ((159.865578 -11.168170, 159.865578 -39.834278, 111.086281 -39.834278, 111.086281 -11.168170, 159.865578 -11.168170)))"^^geo:wktLiteral, ?wkt))
  }
}
```

**Comment by jamiefeiss** [comment]
Resolved in #62
